gapic-generator/lib/gapic/path_pattern/segment.rb

```diff
@@ -16,58 +16,260 @@
+
+      ##
+      # Whether the segment is positional
+      # @return [Boolean]
+      def positional?
+        true
+      end
+
+      ##
+      # Whether the segment provides a resource pattern
+      # @return [Boolean]
+      def resource_pattern?
+        true
+      end
+
+      ##
+      # Whether the segment provides a nontrivial resource pattern
+      # @return [Boolean]
+      def nontrivial_resource_pattern?
+        false
+      end
+
+      ##
+      # Whether the segment provides arguments
+      # @return [Boolean]
+      def provides_arguments?
+        true
+      end
+
+      ##
+      # Names of the segment's arguments
+      # @return [Array<String>]
+      def arguments
+        [position.to_s]
+      end
+
+      ##
+      # Returns a segment's pattern filled with dummy values
+      #   names of the values are generated starting from the index provided
+      # @param start_index [Integer] a starting index for dummy value generation
+      # @return [String] a pattern filled with dummy values
+      def expected_path_for_dummy_values start_index
+        "value#{start_index}"
+      end
+
+      ##
+      # Path string for this segment
+      # @return [String]
+      def path_string
+        "\#{#{position}}"
+      end
+
+      ##
+      # A pattern template for this segment
+      # @return [String]
+      def pattern_template
+        pattern
+      end
+
+      # @private
+      def == other
+        return false unless other.is_a? self.class
+
+        (pattern == other.pattern && position == other.position)
+      end
+    end
+
+    # A ResourceId segment in a path pattern.
+    #  ResourceId segments can be simple, with one resource name
+    #  or complex, with multiple resource names divided by separators
     #
-    # @!attribute [r] name
-    #   @return [String, Integer] The name of a named segment, or the position
-    #     of a positional segment.
+    # @!attribure [r] type
+    #   @return [String] The type of this segment
     # @!attribute [r] pattern
-    #   @return [String, nil] The pattern of the segment, nil if not set.
-    class Segment
-      attr_reader :name, :pattern
+    #   @return [String] The pattern of the segment, for the positional segment it is also
+    #     a pattern of its resource
+    # @!attribute [r] resource_names
+    #   @return [Array<String>] The resource names in this segment
+    # @!attribute [r] resource_patterns
+    #   @return [Array<String>] The resource patterns associated with
+    #     the resource_names of this segment
+    class ResourceIdSegment
+      attr_reader :type, :pattern, :resource_names, :resource_patterns
 
-      def initialize name, pattern
-        @name    = name
-        @pattern = pattern
+      def initialize type, pattern, resource_names, resource_patterns = []
+        @type              = type
+        @pattern           = pattern
+        @resource_names    = resource_names
+        @resource_patterns = resource_patterns || []
```

Comment:
Don't need the fallback here since the parameter default is already the empty array. Unless you expect callers to pass `nil` explicitly.